src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -210,3 +236,248 @@ g.test('multisampled_copy_restrictions')
+      isSuccess
+    );
+  });
+
+g.test('copy_ranges')
+  .params(
+    params()
+      .combine(
+        poptions('copyRectangleOffsets', [
+          { x: 0, y: 0, z: 0, width: 0, height: 0, depth: 1 },
+          { x: 4, y: 0, z: 0, width: 0, height: 0, depth: 1 },
+          { x: 8, y: 0, z: 0, width: 0, height: 0, depth: 1 },
+          { x: 0, y: 2, z: 0, width: 0, height: 0, depth: 1 },
+          { x: 0, y: 4, z: 0, width: 0, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 0, width: -4, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 0, width: -8, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 0, width: 0, height: -2, depth: 1 },
+          { x: 0, y: 0, z: 0, width: 0, height: -4, depth: 1 },
+          { x: 0, y: 0, z: 0, width: 4, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 0, width: 0, height: 2, depth: 1 },
+          { x: 0, y: 0, z: 1, width: 0, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 2, width: 0, height: 0, depth: 1 },
+          { x: 0, y: 0, z: 0, width: 0, height: 0, depth: 2 },
+          { x: 0, y: 0, z: 1, width: 0, height: 0, depth: 2 },
+        ])
+      )
+      .combine(poptions('srcCopyLevel', [0, 1]))
+      .combine(poptions('dstCopyLevel', [0, 1]))
+  )
+  .fn(async t => {
+    const { copyRectangleOffsets, srcCopyLevel, dstCopyLevel } = t.params;
+
+    const kTextureWidth = 16;
+    const kTextureHeight = 8;
+    const kMipLevelCount = 2;
+    const kArrayLayerCount = 2;
+
+    const srcTexture = t.device.createTexture({
+      size: { width: kTextureWidth, height: kTextureHeight, depth: kArrayLayerCount },
+      format: 'rgba8unorm',
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_SRC,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: kTextureWidth, height: kTextureHeight, depth: kArrayLayerCount },
+      format: 'rgba8unorm',
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_DST,
+    });
+
+    const srcTextureWidthAtLevel = kTextureWidth >> srcCopyLevel;
+    const srcTextureHeightAtLevel = kTextureHeight >> srcCopyLevel;
+    const dstTextureWidthAtLevel = kTextureWidth >> dstCopyLevel;
+    const dstTextureHeightAtLevel = kTextureHeight >> dstCopyLevel;
+
+    const copyOrigin = { x: copyRectangleOffsets.x, y: copyRectangleOffsets.y, z: 0 };
+
+    const copyWidth =
+      Math.min(srcTextureWidthAtLevel, dstTextureWidthAtLevel) +
+      copyRectangleOffsets.width -
+      copyOrigin.x;
+    const copyHeight =
+      Math.min(srcTextureHeightAtLevel, dstTextureHeightAtLevel) +
+      copyRectangleOffsets.height -
+      copyOrigin.y;
+
+    {
+      const isSuccess =
+        copyWidth <= srcTextureWidthAtLevel &&
+        copyHeight <= srcTextureHeightAtLevel &&
+        copyOrigin.x + copyWidth <= dstTextureWidthAtLevel &&
+        copyOrigin.y + copyHeight <= dstTextureHeightAtLevel &&
+        copyOrigin.z + copyRectangleOffsets.depth <= kArrayLayerCount;
```

Comment:
nit: the semantic of copyBoxOffsets.depth seems a bit different from width and height, maybe we could make it the same?

Reply:
Done